DOCS-719: TPC updates (#1026)
hazelcast/hz-docs: docs/modules/cluster-performance/pages/thread-per-core-tpc.adoc

= Thread-Per-Core (TPC)
:description: Find out more about Thread-Per-Core (TPC) and how to enable this feature on clients and cluster members.
:page-enterprise: true

> You can enable Thread-Per-Core (TPC) on clients and cluster members to improve system performance, especially on machines with either fewer or more CPU cores than the recommended hardware listed on the xref:ROOT:production-checklist.adoc[]. 

Author: ```suggestion
Enable Thread-Per-Core (TPC) on both clients and cluster members to improve system performance, especially on machines with either fewer or more CPU cores than the recommended hardware listed on the xref:ROOT:production-checklist.adoc[]. 
```

SEDA can lead to throughput and latency issues on machines with limited cores (fewer than 8), as the frequent context switching and state sharing can add significant overhead. Conversely, on machines with a very high number of CPUs (more than 32), a bottleneck can be created by the limited number of I/O threads vs. operation threads. Typically, smaller servers are more readily available through virtualization and for cloud-specific workloads, which can limit system performance in a SEDA environment. 

A Thread-Per-Core (TPC) design uses one thread for every CPU core and every thread does everything. With each TPC thread able to do networking/storage and compute, scaling is mostly a case of increasing the number of threads to spread the work. This means you can scale up and down easily based on the number of available cores, and make more efficient use of resources on machines with either a low or high number of cores. Every task is non-blocking and you can scale vertically much easier, which means high throughput, low latency and significantly improved system performance.  

> When TPC is enabled, smart clients will connect to the configured number of cores (default is 1) but unisocket clients will establish connections to all cluster members (which can reduce performance in cloud environments). If the server side has also enabled TPC, the client will open connections to all cores of all members and route partition-specific requests to the correct cores.  TPC-enabled servers will continue using the same ports for discovery, which means there's no difference in how the cluster member list is created and TPC-aware servers will be backwards compatible with clients not using TPC.

Reviewer: Uniisocket will completely ignore TPC. Unisocket client will connect to random member in cluster.

Author: So you _must_ use smart clients with TPC?

Author: ```suggestion
When TPC is enabled, smart clients will connect to the configured number of cores (default is 1). If TPC is enabled on both client and server, the client will connect directly to one of the TPC threads instead of using the legacy network threads. TPC-enabled servers will continue using the same ports for discovery, which means there's no difference in how the cluster member list is created and TPC-aware servers will be backwards compatible with clients not using TPC.

NOTE: Unisocket clients will ignore TPC and connect to a random member in the cluster. You must use smart clients with TPC.
```

Reviewer: Correct. TPC is only supported for smart clients (so the non unisocket ones)

> When TPC is enabled, smart clients will connect to the configured number of cores (default is 1) but unisocket clients will establish connections to all cluster members (which can reduce performance in cloud environments). If the server side has also enabled TPC, the client will open connections to all cores of all members and route partition-specific requests to the correct cores.  TPC-enabled servers will continue using the same ports for discovery, which means there's no difference in how the cluster member list is created and TPC-aware servers will be backwards compatible with clients not using TPC.

Reviewer: "If the server side has also enabled TPC, the client will open connections to all cores of all members and route partition-specific requests to the correct cores."

If TPC is enabled on both client and server, by default the client will connect directly to 1 of the tpc threads instead of using the legacy network threads.

Author: > configured number of cores (default is 1)

Where/how is this configured? Is there a client TPC config too? https://github.com/hazelcast/hazelcast/blob/v5.3.6/hazelcast/src/main/resources/hazelcast-client-full-example.xml only has enabled or not. We have eventloop at the server which defaults to 12 but this seems to relate to something else...

Reviewer: In the CLientTpcConfig there is a connection count:

```
public final class ClientTpcConfig {

    private boolean enabled;

    private int connectionCount;```

So if the server has configured e.g. 12 (that should not be a default), the client can pick between -1 and 12.

```

    /**
     * Sets the number of connections to TPC ports offered by a Hazelcast member.
     *
     * <ol>
     *     <li>If set to a negative value, an IllegalArgumentException will be
     *     thrown.</li>
     *     <li>If set to 0, the client will connect to every TPC port.
     *     </li>
     *     <li>If set to the same number as returned by the server, the client
     *     will connect to every TPC port.
     *     </li>
     *     <li>If set to a number larger than 0 and smaller than the number of
     *     returned TPC ports, the client will randomize the list of ports and
     *     make the configured number of connections.
     *     </li>
     *     <li>If set to a number larger than the number of TPC ports, the client
     *     will connect to each tpc port (has same effect as configuring 0)
     *     </li>
     * </ol>
     * Increasing the number of connections leads to more packets with a smaller
     * payload and this can lead to a performance penalty. Also in cloud environments
     * e.g. AWS there can be a packets per second limit (pps) and it pretty easy
     * to run into this limit if an equal number of connections is created as
     * TPC ports on the server and a lot of small interactions are done e.g. a
     * map.get with small payloads.
     *
     * @param connectionCount throws IllegalArgumentException when connectionCount
     *                        is negative.
     * @since 5.4
     */
    public ClientTpcConfig setConnectionCount(int connectionCount) {
        this.connectionCount = checkNotNegative(connectionCount, "connectionCount");
        return this;
    }```

Author: Should we link to the API docs for this then? i.e. https://docs.hazelcast.org/docs/latest/javadoc/com/hazelcast/config/tpc/package-summary.html

I'm not sure if you want to expand the current paragraph we're reviewing or not. Based on what I understand from above I would either remove the (default is 1) comment or expand it and link to the API docs. WDYT?

Reviewer: I would provide an explanation of this property and you can use the javadoc as a source of inspiration.

Author: ```suggestion
When TPC is enabled, smart clients will connect to the number of cores specified in the client config. If TPC is enabled on both client and server, clients connect directly to one of the TPC threads instead of using the legacy network threads. TPC-enabled servers continue to use the same ports for discovery, which means there's no difference in how the cluster member list is created and TPC-aware servers are backward compatible with clients that don't use TPC.

NOTE: Unisocket clients will ignore TPC and connect to a random member in the cluster. You must use smart clients with TPC. For further information on the `ClientTpcConfig` class used to specify the number of cores to use for connections, refer to https://docs.hazelcast.org/docs/{full-version}/javadoc/com/hazelcast/client/config/ClientTpcConfig.html[Class ClientTpcConfig] in the Java API documentation.
```

[[tpc-config]]
== Configuration Options

TPC is disabled by default and must be enabled on each cluster member and client.  

> NOTE: If you configure a client with TPC and the server doesn't support it, the client is ignored. If you configure a client without TPC and the server supports it, TPC is ignored (server and client will still communicate).

Reviewer: 'the client is ignored' --> The client will not make use of TPC.

Author: ```suggestion
NOTE: If you configure a client with TPC and the server doesn't support it, or if you configure a client without TPC and the server supports it, TPC will be ignored in both cases (server and client will still communicate).
```

Reviewer: Yes. If either side doesn't have TPC configured, then the client will not use TPC.

Internal:

If TPC on the server is enabled, the server sends back a list of TPC ports the client can pick from.

If TPC on the server is disabled, this list is empty and the client has nothing to pick from and hence will only use the classic networking.
